lib/src/main/kotlin/tsl/trees/AVLTree.kt

```diff
@@ -0,0 +1,159 @@
+        updateHeight(balancedNode.rightChild)
+        updateHeight(balancedNode)
+
+        return balancedNode
+    }
+
+    private fun rotateRight(oldUpperNode: AVLNode<K, V>): AVLNode<K, V> {
+        val newUpperNode = oldUpperNode.leftChild ?: return oldUpperNode
+        oldUpperNode.leftChild = newUpperNode.rightChild
+        newUpperNode.rightChild = oldUpperNode
+
+        if (root == oldUpperNode) root = newUpperNode // if root was rotated, set new root
+
+        return newUpperNode
+    }
+
+    private fun rotateLeft(oldUpperNode: AVLNode<K, V>): AVLNode<K, V> {
+        val newUpperNode = oldUpperNode.rightChild ?: return oldUpperNode
+        oldUpperNode.rightChild = newUpperNode.leftChild
+        newUpperNode.leftChild = oldUpperNode
+
+        if (root == oldUpperNode) root = newUpperNode // if root was rotated, set new root
+
+        return newUpperNode
+    }
+
+    private fun updateHeight(node: AVLNode<K, V>?): Int {
+        if (node == null) return -1
+        node.height = max(getHeight(node.leftChild), getHeight(node.rightChild)) + 1
+        return node.height
+    }
+
+    private fun getBalanceFactor(node: AVLNode<K, V>): Int =
+        getHeight(node.rightChild) - getHeight(node.leftChild)
+
+    private fun getHeight(node: AVLNode<K, V>?): Int = node?.height ?: -1
+
+    public override fun delete(key: K): V? {
+        val deletedValue: V =
+            search(key) ?: return null // if key isn't in the tree, there's nothing to delete
+
+        deleteAndBalanceRecursively(root, key)
+
+        return deletedValue
+    }
+
+    private fun deleteAndBalanceRecursively(
+        currNode: AVLNode<K, V>?,
+        keyToDelete: K
+    ): AVLNode<K, V>? {
+
+        when {
+            currNode == null -> return null // node to be deleted was not found
+            keyToDelete < currNode.key -> // node to be deleted is in the left subtree
+            currNode.leftChild = deleteAndBalanceRecursively(currNode.leftChild, keyToDelete)
+            keyToDelete > currNode.key -> // node to be deleted is in the right subtree
+            currNode.rightChild = deleteAndBalanceRecursively(currNode.rightChild, keyToDelete)
+            else -> {
+                if (currNode.leftChild == null || currNode.rightChild == null) {
+                    return currNode.leftChild ?: currNode.rightChild
+                } else {
+                    val successor = getMinNodeRecursively(currNode.rightChild)
+                    if (successor != null) {
+                        // copy its successor
+                        currNode.key = successor.key
+                        currNode.value = successor.value
+
+                        // delete original successor node from tree
+                        val newSubtree =
+                            deleteAndBalanceRecursively(currNode.rightChild, successor.key)
+                        if (newSubtree != null) currNode.rightChild = newSubtree

```

Comment:
```suggestion
                        currNode.rightChild = newSubtree
```
в этой ветке у нас 1-е. ключ узла совпадает с ключом, по которому нужно удалить 2-е. есть левый и правый узел. т.е. всегда есть на что заменить. теперь, когда создаем newSubtree, в нем может быть сразу тот нод на который заменяем, мы его удалим и получим null. удалили нод и получили пустую правую ветку. почему-то, если она пустая мы ее не присваиваем, а просто оставляем не удаленное до конца дерево в правой ветки. короче, тут не должно быть ифа